pinot-common/src/main/java/org/apache/pinot/common/utils/request/RequestUtils.java

```diff
@@ -664,4 +670,49 @@ public static void applyTimestampIndexOverrideHints(
       }
     }
   }
+
+  /**
+   * Infers the expression type by recursively traversing the expression tree for function calls. The provided schema
+   * is used to resolve the types for identifier expressions. Note that for function calls, only scalar functions are
+   * supported here currently. Transform functions that don't have equivalent scalar functions, and aggregation
+   * functions aren't supported and {@link ColumnDataType#UNKNOWN} will be returned for them.
+   */
+  @Nullable
+  public static ColumnDataType inferExpressionType(@Nullable Expression expression, Schema schema) {
```

Comment:
I don't think we want to accept `null` expression here. If we don't accept `null` expression, the return is also always not null